src/js/github.js

```diff
@@ -0,0 +1,123 @@
+		var reset = new Date(1000 * window.parseInt(githubHTTPClient.getResponseHeader("x-ratelimit-reset")));
+		console.log("Rate limit used " + used + "/" + limit + " (resets " + reset.toISOString() + ")");
+
+		var result = JSON.parse(githubHTTPClient.responseText);
+		if (githubHTTPClient.status===403) {
+			if (!github.isSignedIn() && (result.message.indexOf("rate limit") !== -1)) {
+				done(null, "Exceeded GitHub rate limits. Try signing in from the editor.");
+			} else {
+				done(null, result.message);
+			}
+		} else if (githubHTTPClient.status===401) {
+			github.signOut();
+			done(null, "Authorization check failed.  Try reloading or signing back in from the editor.");
+		} else if (githubHTTPClient.status>=500) {
+			done(null, "HTTP Error "+ githubHTTPClient.status + " - " + githubHTTPClient.statusText + ".");
+		} else if (githubHTTPClient.status!==200 && githubHTTPClient.status!==201) {
+			done(null, "HTTP Error "+ githubHTTPClient.status + " - " + githubHTTPClient.statusText);
+		} else {
+			var result = JSON.parse(githubHTTPClient.responseText);
+			var code=result["files"]["script.txt"]["content"];
+			done(code, null);
+		}
+	}
+
+	if (github.isSignedIn()) {
+		var oauthAccessToken = storage_get("oauth_access_token");
+		githubHTTPClient.setRequestHeader("Authorization", "Token "+oauthAccessToken);
+	}
+	githubHTTPClient.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
+	githubHTTPClient.send();
+};
+
+github.save = function(title, code, done) {
+	var oauthAccessToken = storage_get("oauth_access_token");
+	if (typeof oauthAccessToken !== "string") {
+		printUnauthorized();
+		return;
+	}
+
+	var gistToCreate = {
+		"description" : title,
+		"public" : true,
+		"files": {
+			"readme.txt" : {
+				"content": "Play this game by pasting the script in http://www.puzzlescript.net/editor.html"
+			},
+			"script.txt" : {
+				"content": code
+			}
+		}
+	};
+
+	var githubURL = "https://api.github.com/gists";
+	var githubHTTPClient = new XMLHttpRequest();
+	githubHTTPClient.open("POST", githubURL);
+	githubHTTPClient.onreadystatechange = function() {		
+		if(githubHTTPClient.readyState!=4) {
+			return;
+		}		
+		var result = JSON.parse(githubHTTPClient.responseText);
+		if (githubHTTPClient.status===403) {
+			done(null, result.message);
+		} else if (githubHTTPClient.status===401) {
+			github.signOut();
+			done(null, "Authorization check failed.  You have to log back into GitHub (or give it permission again or something).");
+		} else if (githubHTTPClient.status>=500) {
+			done(null, "HTTP Error "+ githubHTTPClient.status + " - " + githubHTTPClient.statusText + ".");
+		} else if (githubHTTPClient.status!==200 && githubHTTPClient.status!==201) {
+			github.signOut();
+			done(null, "HTTP Error "+ githubHTTPClient.status + " - " + githubHTTPClient.statusText + ".  Try giving puzzlescript permission again, that might fix things...");
+		} else {
+			done(result.id, null);
+		}
```

Comment:
nit: This block looks similar to the one in `load()` -- possibly worth extracting a helper like `handleGithubResponse()`.

Reply:
I don't think multiplexing http error codes as function parameters is beneficial complexity wise, but a very fair suggestion